### PR TITLE
JUCX: tag_probe message functionality.

### PR DESCRIPTION
**Pull request by petro-rudenko**
## What
Ucp tagProbe message functionality: `tagProbe` and `recvTagMessage`

## Why ?
As requested in #4723 and also for SparkUCX to submit recv request only when there's a message in recv queue.

### CODE REVIEW

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 22-Feb-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/755d64881727513e5417b25bd71156dba1393c95/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/a0be07bfaa4fad6226912a067c76cb19f3986d7f/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/acd39f70e842232736056867990faef6cd7b9ac2/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/90edb844d9b036889da301727d961b373e89b168/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/a1015982fd84147a05a313be777c2d5e124b4c77/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/6dd629c2446804ca585584d2b6dfd8bae0565f23/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/c504155ff6d1633c3f804905f3c5946bb572ebaa/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/ff221a4d62749b4039c477fc0004cf5ba2a7a9de/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/5c195d3a075e1e614b065bd32b97829a157d1f90/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/87dff736b9299218bf962142c14f6b41a841941e/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/4e17bdf8d0f2d9d2376e8870d6fd2e6b80236ec3/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/6236b5099224e0ff7a8576dc0276374bb296026e/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/30a75b7742f5d88b368642acd8c09704040a3294/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/b685882a2823cb74ec32e77c4e73babe6e6ecf9f/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/65e1eeae9ee6264d834f5bbb037d55053e2bca22/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/c796a678389d8f540a14be8a5821fd651c352b6f/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/3fdeb68506f64f2e0bcd2d3305415bad6eb64770/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/601d335e89e11034d965e3b44def7d903b9ada25/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/4a6b59dcf013905ecce96366a4f6990818566ca1/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/0bb4c9e56e06e5e2dff1b84135187eee9c5cd629/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/f29d706c7b2cf13bc669a5ff1985067ea4f74ad2/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/60ed9f8ffddbaaf9b935ab50990f7e5da0fd3cc7/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/9a04b0f39f03cc6cf6c024ce75c763db09338282/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/119c65ed81d1c916f3a0242a7a3c145ea1665b65/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/dc77f903c785a4c94d1ca10eb1221135/raw/180edb4435545570c4072f2bc429a1f892a4d252/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

File: bindings/java/src/main/native/worker.cc [hunk 1]
```diff
@@ -152,7 +152,44 @@ Java_org_openucx_jucx_ucp_UcpWorker_recvTaggedNonBlockingNative(JNIEnv *env, jcl
...
+                                                                       jlong msg_ptr,
+                                                                       jobject callback)
+{
+    ucs_status_ptr_t request = ucp_tag_msg_recv_nb((ucp_worker_h)ucp_worker_ptr,
+                                                   (void *)laddr, size,
+                                                   ucp_dt_make_contig(1),
+                                                   (ucp_tag_message_h)msg_ptr,
+                                                   recv_callback);
+
+    ucs_trace_req("JUCX: Tag_msg_recv_nb request %p, msg size: %zu, msg: %p", request, size,
```
Comment (yosefe) [inline bindings/java/src/main/native/worker.cc]: tag_msg_recv_nb

File: bindings/java/src/main/native/worker.cc [hunk 2]
```diff
@@ -152,7 +152,44 @@ Java_org_openucx_jucx_ucp_UcpWorker_recvTaggedNonBlockingNative(JNIEnv *env, jcl
...
+                                                              jlong ucp_worker_ptr,
+                                                              jlong tag, jlong tag_mask,
+                                                              jboolean remove)
+{
+    ucp_tag_recv_info_t info_tag;
+    ucp_tag_message_h msg_tag = ucp_tag_probe_nb((ucp_worker_h)ucp_worker_ptr, tag, tag_mask,
+                                                 remove, &info_tag);
+    jobject result = NULL;
+
+    if (msg_tag != NULL) {
```
Comment (yosefe) [inline bindings/java/src/main/native/worker.cc]: also check for UCS_PTR_IS_ERR(_ptr)

Comment (petro-rudenko) [reply bindings/java/src/main/native/worker.cc]: https://github.com/openucx/ucx/blob/master/src/ucp/api/ucp.h#L2650-L2652:

```
 @return NULL                      - No match found.
 * @return Message handle (not NULL) - If message is matched the message handle
 *                                     is returned.
```

Comment (yosefe) [reply bindings/java/src/main/native/worker.cc]: ok

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 24-Feb-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/7e607d26ecf3ffd1aad58b4e900a1041575d7d29/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/3548a7c278e5a7b2ac6bd67eaea2fb0317281cf6/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/6c31ebeb75bd5be25f8cfdbec435781b328f772f/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/36a5e8b94378eb9d4d28d6c0c8cf7e131d8657b1/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/dfd06dfbe15437af78360aae5f17cf109421c3ee/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/70490dbe49898256c42b728fdca36f20d5a94bf6/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/369456c8c833cf82b6fb6fbfc371490cdd55488f/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/c3046e3025c3fd51d181f738f805439bb6244c36/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/938e682486a032c3a6410f9c9439ac8849ee912d/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/26462e7c0f0dbcce6572c2107bafe925a047fcd6/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/4e6fa4b5d0d51652d1a4bda29207fc48bc5d1a0d/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/c428e428276b0ff7c711e7258d452b6c0994805f/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/5431c676a47a7f68493ec76b259314267436cfcf/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/e702c0383cea8b447eb3f1c5453b0bc5e70d1e4f/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/60384388cb13d73eb3ee89d151b146392edd2521/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/86ee0d457bf0c0fb83be29f6b88ace627f16d561/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/858a1b17d868d095ae69acce30808e1c53ef60ae/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/e9ff770f4724e40ec8e4bd7579ca4e21dcf30a5d/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/5369dc317794ab0e5271302c3fde870306a778fa/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/14baff615ece40c2ae63ee2a2c0cca3d8c244714/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/6da413e2f415bb38d4cac476c5d9f9a35246d055/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/1f0b51f1375e4aa1b18c60cfaf3416e1db99c4e7/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/3516e22d74cfee1e07e9538b082b6665a2e4e643/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/7e4a5d932722e2399c283e740f4c3f578c1d994b/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/7688510a2dbd941345e271b2795757ff/raw/14d3798b91a737750c559918e9a833b630f17c71/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

**Pull request merged**